integration-tests/tests/client/sharding_upgrade.rs

```diff
@@ -9,21 +7,202 @@ use near_logger_utils::init_test_logger;
+            for tx in txs {
+                for j in 0..self.num_validators {
+                    env.clients[j].process_tx(tx.clone(), false, false);
+                }
+            }
+        }
+
+        // produce block
+        let block_producer = {
+            let epoch_id = env.clients[0]
+                .runtime_adapter
+                .get_epoch_id_from_prev_block(&head.last_block_hash)
+                .unwrap();
+            env.clients[0].runtime_adapter.get_block_producer(&epoch_id, height).unwrap()
+        };
+        let block_producer_client = env.client(&block_producer);
+        let mut block = block_producer_client.produce_block(height).unwrap().unwrap();
+        set_block_protocol_version(
+            &mut block,
+            block_producer.clone(),
+            SIMPLE_NIGHTSHADE_PROTOCOL_VERSION,
+        );
+        // process block, this also triggers chunk producers for the next block to produce chunks
+        for j in 0..self.num_clients {
+            env.process_block(j as usize, block.clone(), Provenance::NONE);
+        }
+
+        env.process_partial_encoded_chunks();
+
+        // after state split, check chunk extra exists and the states are correct
+        for account_id in self.initial_accounts.iter() {
+            check_account(env, account_id, &block);
+        }
+    }
+
+    /// check that all accounts in `accounts` exist in the current state
+    fn check_accounts(&mut self, accounts: &[AccountId]) {
+        let head = self.env.clients[0].chain.head().unwrap();
+        let block = self.env.clients[0].chain.get_block(&head.last_block_hash).unwrap().clone();
+        for account_id in accounts {
+            check_account(&mut self.env, account_id, &block)
+        }
+    }
+
+    /// This functions checks that the outcomes of all transactions and associated receipts
+    /// have successful status
+    fn check_tx_outcomes(&mut self) {
+        let env = &mut self.env;
+        let head = env.clients[0].chain.head().unwrap();
+        let block = env.clients[0].chain.get_block(&head.last_block_hash).unwrap().clone();
+        // check execution outcomes
+        let shard_layout = env.clients[0]
+            .runtime_adapter
+            .get_shard_layout_from_prev_block(&head.last_block_hash)
+            .unwrap();
+        let mut txs: Vec<_> = self.txs_by_height.values().flatten().collect();
+        txs.extend(&self.init_txs);
+
+        for tx in txs {
+            let id = &tx.get_hash();
+            let account_id = &tx.transaction.signer_id;
+            let shard_uid = account_id_to_shard_uid(account_id, &shard_layout);
+            for (i, account_id) in env.validators.iter().enumerate() {
+                if env.clients[i].runtime_adapter.cares_about_shard(
+                    Some(account_id),
+                    block.header().prev_hash(),
+                    shard_uid.shard_id(),
+                    true,
+                ) {
```

Comment:
nit: multiline conditional are hard to read, it's a good place to extract a variable.